Compilation exception in mapped type

<!-- BUGS: Please use this template. -->
<!-- QUESTIONS: This is not a general support forum! Ask Qs at http://stackoverflow.com/questions/tagged/typescript -->
<!-- SUGGESTIONS: See https://github.com/Microsoft/TypeScript-wiki/blob/master/Writing-Good-Design-Proposals.md -->

**TypeScript Version:**  nightly (2.2.0-dev.20161219)

**Code**

```ts
type Foo<T, F extends keyof T> = {
    pf: {[P in F]?: T[P]},
    pt: {[P in T]?: T[P]}, // note: should be in keyof T
};
type O = {x: number, y: boolean};
let o: O = {x: 5, y: false};
let f: Foo<O, 'x'> = {
    pf: {x: 7},
    pt: {x: 7, y: false},
};
```

**Expected behavior:**

Compiling only the first type declaration produces the following expected error:
```
$ node node_modules/typescript/bin/tsc test.ts
test.ts(4,16): error TS2322: Type '{}' is not assignable to type 'string'.
test.ts(4,21): error TS2536: Type 'P' cannot be used to index type 'T'.
```

**Actual behavior:**
Compiling the whole example with a clean nightly install without `tsconfig` produces:
```
$ node node_modules/typescript/bin/tsc test.ts
C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:49834
                throw e;
                ^

TypeError: Cannot read property 'flags' of undefined
    at getModifiersTypeFromMappedType (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:23125:60)
    at resolveMappedTypeMembers (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:23068:49)
    at resolveStructuredTypeMembers (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:23153:25)
    at getPropertyOfType (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:23307:32)
    at C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:28345:47
    at applyToContextualType (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:28321:24)
    at getTypeOfPropertyOfContextualType (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:28344:20)
    at getContextualTypeForObjectLiteralElement (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:28368:40)
    at getContextualType (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:28442:28)
    at checkExpressionForMutableLocation (C:\Users\...\Documents\TCM\git\tmp\node_modules\typescript\lib\tsc.js:31213:69)
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -4612,8 +4612,8 @@ namespace ts {
                     // the modifiers type is T. Otherwise, the modifiers type is {}.
                     const declaredType = <MappedType>getTypeFromMappedTypeNode(type.declaration);
                     const constraint = getConstraintTypeFromMappedType(declaredType);
-                    const extendedConstraint = constraint.flags & TypeFlags.TypeParameter ? getConstraintOfTypeParameter(<TypeParameter>constraint) : constraint;
-                    type.modifiersType = extendedConstraint.flags & TypeFlags.Index ? instantiateType((<IndexType>extendedConstraint).type, type.mapper || identityMapper) : emptyObjectType;
+                    const extendedConstraint = constraint && constraint.flags & TypeFlags.TypeParameter ? getConstraintOfTypeParameter(<TypeParameter>constraint) : constraint;
+                    type.modifiersType = extendedConstraint && extendedConstraint.flags & TypeFlags.Index ? instantiateType((<IndexType>extendedConstraint).type, type.mapper || identityMapper) : emptyObjectType;
                 }
             }
             return type.modifiersType;

--- a/tests/baselines/reference/mappedTypeErrors.errors.txt
+++ b/tests/baselines/reference/mappedTypeErrors.errors.txt
@@ -45,9 +45,11 @@ tests/cases/conformance/types/mapped/mappedTypeErrors.ts(130,5): error TS2322: T
 tests/cases/conformance/types/mapped/mappedTypeErrors.ts(131,5): error TS2322: Type '{ a: string; }' is not assignable to type '{ [x: string]: any; a?: number | undefined; }'.
   Types of property 'a' are incompatible.
     Type 'string' is not assignable to type 'number | undefined'.
+tests/cases/conformance/types/mapped/mappedTypeErrors.ts(137,16): error TS2322: Type '{}' is not assignable to type 'string'.
+tests/cases/conformance/types/mapped/mappedTypeErrors.ts(137,21): error TS2536: Type 'P' cannot be used to index type 'T'.
 
 
-==== tests/cases/conformance/types/mapped/mappedTypeErrors.ts (24 errors) ====
+==== tests/cases/conformance/types/mapped/mappedTypeErrors.ts (26 errors) ====
     
     interface Shape {
         name: string;
@@ -250,3 +252,21 @@ tests/cases/conformance/types/mapped/mappedTypeErrors.ts(131,5): error TS2322: T
 !!! error TS2322: Type '{ a: string; }' is not assignable to type '{ [x: string]: any; a?: number | undefined; }'.
 !!! error TS2322:   Types of property 'a' are incompatible.
 !!! error TS2322:     Type 'string' is not assignable to type 'number | undefined'.
+    
+    // Repro from #13044
+    
+    type Foo2<T, F extends keyof T> = {
+        pf: {[P in F]?: T[P]},
+        pt: {[P in T]?: T[P]}, // note: should be in keyof T
+                   ~
+!!! error TS2322: Type '{}' is not assignable to type 'string'.
+                        ~~~~
+!!! error TS2536: Type 'P' cannot be used to index type 'T'.
+    };
+    type O = {x: number, y: boolean};
+    let o: O = {x: 5, y: false};
+    let f: Foo2<O, 'x'> = {
+        pf: {x: 7},
+        pt: {x: 7, y: false},
+    };
+    

--- a/tests/baselines/reference/mappedTypeErrors.js
+++ b/tests/baselines/reference/mappedTypeErrors.js
@@ -131,6 +131,20 @@ let x1: T2 = { a: 'no' };  // Error
 let x2: Partial<T2> = { a: 'no' }; // Error
 let x3: { [P in keyof T2]: T2[P]} = { a: 'no' };  // Error
 
+// Repro from #13044
+
+type Foo2<T, F extends keyof T> = {
+    pf: {[P in F]?: T[P]},
+    pt: {[P in T]?: T[P]}, // note: should be in keyof T
+};
+type O = {x: number, y: boolean};
+let o: O = {x: 5, y: false};
+let f: Foo2<O, 'x'> = {
+    pf: {x: 7},
+    pt: {x: 7, y: false},
+};
+
+
 //// [mappedTypeErrors.js]
 function f1(x) {
     var y; // Error
@@ -204,6 +218,11 @@ c.setState({ c: true }); // Error
 var x1 = { a: 'no' }; // Error
 var x2 = { a: 'no' }; // Error
 var x3 = { a: 'no' }; // Error
+var o = { x: 5, y: false };
+var f = {
+    pf: { x: 7 },
+    pt: { x: 7, y: false }
+};
 
 
 //// [mappedTypeErrors.d.ts]
@@ -268,3 +287,17 @@ declare let x2: Partial<T2>;
 declare let x3: {
     [P in keyof T2]: T2[P];
 };
+declare type Foo2<T, F extends keyof T> = {
+    pf: {
+        [P in F]?: T[P];
+    };
+    pt: {
+        [P in T]?: T[P];
+    };
+};
+declare type O = {
+    x: number;
+    y: boolean;
+};
+declare let o: O;
+declare let f: Foo2<O, 'x'>;

--- a/tests/cases/conformance/types/mapped/mappedTypeErrors.ts
+++ b/tests/cases/conformance/types/mapped/mappedTypeErrors.ts
@@ -131,3 +131,16 @@ type T2 = { a?: number, [key: string]: any };
 let x1: T2 = { a: 'no' };  // Error
 let x2: Partial<T2> = { a: 'no' }; // Error
 let x3: { [P in keyof T2]: T2[P]} = { a: 'no' };  // Error
+
+// Repro from #13044
+
+type Foo2<T, F extends keyof T> = {
+    pf: {[P in F]?: T[P]},
+    pt: {[P in T]?: T[P]}, // note: should be in keyof T
+};
+type O = {x: number, y: boolean};
+let o: O = {x: 5, y: false};
+let f: Foo2<O, 'x'> = {
+    pf: {x: 7},
+    pt: {x: 7, y: false},
+};

